agent/connect/ca/provider_consul.go

```diff
@@ -170,10 +170,126 @@ func (c *ConsulProvider) GenerateRoot() error {
 	return nil
 }
 
+// GenerateIntermediateCSR creates a private key and generates a CSR
+// for another datacenter's root to sign.
+func (c *ConsulProvider) GenerateIntermediateCSR() (string, error) {
+	_, providerState, err := c.getState()
+	if err != nil {
+		return "", err
+	}
+
+	if c.isRoot {
+		return "", fmt.Errorf("provider is the root certificate authority, " +
+			"cannot generate an intermediate CSR")
+	}
+
+	// Create a new private key and CSR.
+	signer, pk, err := connect.GeneratePrivateKey()
+	if err != nil {
+		return "", err
+	}
+
+	csr, err := connect.CreateCSR(c.spiffeID, signer)
+	if err != nil {
+		return "", err
+	}
+
+	// Write the new provider state to the store.
+	newState := *providerState
+	newState.PrivateKey = pk
+	args := &structs.CARequest{
+		Op:            structs.CAOpSetProviderState,
+		ProviderState: &newState,
+	}
+	if err := c.Delegate.ApplyCARequest(args); err != nil {
+		return "", err
+	}
+
+	return csr, nil
+}
+
+// SetIntermediate validates that the given intermediate is for the right private key
+// and writes the given intermediate and root certificates to the state.
+func (c *ConsulProvider) SetIntermediate(intermediatePEM, rootPEM string) error {
+	_, providerState, err := c.getState()
+	if err != nil {
+		return err
+	}
+
+	if c.isRoot {
+		return fmt.Errorf("cannot set an intermediate using another root in the primary datacenter")
+	}
+
+	// Get the key from the incoming intermediate cert so we can compare it
+	// to the currently stored key.
+	intermediate, err := connect.ParseCert(intermediatePEM)
+	if err != nil {
+		return fmt.Errorf("error parsing intermediate PEM: %v", err)
+	}
+	privKey, err := connect.ParseSigner(providerState.PrivateKey)
+	if err != nil {
+		return err
+	}
+
+	// Compare the two keys to make sure they match.
+	b1, err := x509.MarshalPKIXPublicKey(intermediate.PublicKey)
+	if err != nil {
+		return err
+	}
+	b2, err := x509.MarshalPKIXPublicKey(privKey.Public())
+	if err != nil {
+		return err
+	}
+	if !bytes.Equal(b1, b2) {
+		return fmt.Errorf("intermediate cert is for a different private key")
+	}
+
+	// Validate the remaining fields and make sure the intermediate validates against
+	// the given root cert.
+	if !intermediate.IsCA {
+		return fmt.Errorf("intermediate is not a CA certificate")
+	}
+	if got, want := intermediate.URIs[0].String(), c.spiffeID.URI().String(); got != want {
```

Comment:
This could panic - we should check there actually IS a URI before referencing one.

I suggest we at the same time validate there is _exactly_ one URI since we assume that anyway.